File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/ByteEntityTransform.java
package com.loudsight.meta.serialization.transform;

import java.util.Iterator;
import java.util.List;

import com.loudsight.meta.serialization.EntityTransform;
import com.loudsight.meta.serialization.EntityType;

/**
 * EntityTransform implementation for Byte types.
 */
public class ByteEntityTransform extends EntityTransform<Byte> {

    /**
     * Holder for singleton instance.
     */
    private static class ByteEntityTransformHolder {
        private static final ByteEntityTransform INSTANCE = new ByteEntityTransform();
    }
    /**
     * Returns singleton instance of ByteEntityTransform.
     * @return the singleton instance
     */
    public static ByteEntityTransform getInstance() {
        return ByteEntityTransform.ByteEntityTransformHolder.INSTANCE;
    }
    private ByteEntityTransform() {
        super(EntityType.BYTE, Byte.class);
    }


    @Override public void serializeEntity( Byte entity, List<Byte> bytes) {
        bytes.add(EntityType.BYTE.getCode());
        bytes.add(entity);
    }

    @Override public Byte  deserializeEntity(Iterator<Byte> bytes) {
        return bytes.next();
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/transform/ByteEntityTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.serialization.EntityTransform
import com.loudsight.meta.serialization.EntityType

object ByteEntityTransform : EntityTransform<Byte>(EntityType.BYTE, Byte::class) {

    override fun serializeEntity(entity: Byte, bytes: MutableList<Byte>) {
        bytes.add(EntityType.BYTE.code)
        bytes.add(entity)
    }

    override fun deserializeEntity(bytes: Iterator<Byte>): Byte {
        return bytes.next()
    }

    override fun canTransform(entity: Any): Boolean {
        return entity is Byte
    }
}